cxx/pclean/pclean_lib_test.cc

```diff
@@ -0,0 +1,57 @@
+#define BOOST_TEST_MODULE test pclean_csv
+
+#include "pclean/pclean_lib.hh"
+#include <sstream>
+#include <boost/test/included/unit_test.hpp>
+namespace tt = boost::test_tools;
+
+BOOST_AUTO_TEST_CASE(test_translate_observations) {
+  std::stringstream ss(R"""(Column1,Room Type,Monthly Rent,County,State
+0,studio,,Mahoning County,OH
+1,4br,2152.0,,NV
+2,1br,1267.0,Gwinnett County,
+)""");
+
+  DataFrame df = DataFrame::from_csv(ss);
+
+  std::map<std::string, std::string> state_params = {{"strings", "AL AK AZ AR CA CO CT DE DC FL GA HI ID IL IN IA KS KY LA ME MD MA MI MN MS MO MT NE NV NH NJ NM NY NC ND OH OK OR PA RI SC SD TN TX UT VT VA WA WV WI WY"}};
+  std::map<std::string, std::string> br_params = {{"strings", "1br 2br 3br 4br studio"}};
+
+  T_schema schema = {
+    {"County:name",
+      T_clean_relation{{"County"}, false, DistributionSpec("bigram")}},
+    {"County:state",
+      T_clean_relation{{"County"}, false, DistributionSpec("stringcat", state_params)}},
+    {"Room Type",
+      T_clean_relation{{"Obs"}, false, DistributionSpec("stringcat", br_params)}},
+    {"Monthly Rent",
+      T_clean_relation{{"Obs"}, false, DistributionSpec("normal")}},
+    {"County",
+      T_noisy_relation{{"County", "Obs"}, false, EmissionSpec("bigram"), "County:name"}},
+    {"State",
+      T_noisy_relation{{"County", "Obs"}, false, EmissionSpec("bigram"), "County:state"}}};
```

Comment:
For some of the `T_noisy_relation`s, `is_observed` should be `true`, right?

Reply:
translate_observations doesn't use that field of the schema at all, so I don't think it matters at all for the purposes of this test.

Reply:
I think it matters a little for self-documentation, so I'd prefer to set it to `true` just for readability/consistency. In the  HIRM data-ingesting code, these are set to `true` after observations of this relation are encountered.

Reply:
Done.

Reply:
Thanks (sorry for the back and forth, there's a lot going on in this model that's subtle but I think we're getting towards a shared understanding).

It looks to me like the HIRM schema in this test is implementing the following PClean-like schema, is that right?
```
class County
  name ~ bigram
  state ~ stringcat

Obs
  county ~ County
  rent ~ normal
  room ~ stringcat
```
Is this what `schema_helper.make_hirm_schema()` would output? If so, it doesn't quite conform to how I understand Model 5 (and I missed this in earlier code reviews). I would expect the HIRM schema from the above PClean schema to look like this:

```
  T_schema schema = {
    {"Obs:county:name",
      T_clean_relation{{"dCounty"}, false, DistributionSpec("bigram")}},
    {"Obs:county:state",
      T_clean_relation{{"dCounty"}, false, DistributionSpec("stringcat", state_params)}},
    {"Obs:room type",
      T_clean_relation{{"dObs"}, false, DistributionSpec("stringcat", br_params)}},
    {"Obs:monthly rent",
      T_clean_relation{{"dObs"}, false, DistributionSpec("normal")}},
    {"County",
      T_noisy_relation{{"dCounty", "dObs"}, false, EmissionSpec("bigram"), "Obs:county:name"}},
    {"State",
      T_noisy_relation{{"dCounty", "dObs"}, false, EmissionSpec("bigram"), "Obs:county:state"}},
    {"Room Type",
      T_noisy_relation{{"dObs"}, false, DistributionSpec("stringcat", br_params), "Obs:room type}},
    {"Monthly Rent",
      T_noisy_relation{{"dObs"}, false, DistributionSpec("normal"), "Obs:monthly rent}},
};
```

I think this was the source of a lot of my earlier confusion. According to my reading of Model 5, the Obs/Record class should be treated like all of the other latent classes in terms of defining the clean relations. What's special about it is that there's a one-to-one correspondence between its entities and rows of the observed data. The columns of the observed data are all represented by noisy relations, and all of them have "Obs" as an input domain. If you look at the paragraph on page 10 of the Overleaf that begins "Given all this information", I think this is what it implies (in particular, we need an $I_{r_{C.a}}$ for $C = C*$ and the current code omits that).

If you agree, could you make that change in a follow-up PR?

Reply:
Actually, thinking about this more, I think my read is wrong and you're right, that the observations that come directly from $C*$ are clean. No changes are needed here or to `make_hirm_schema()`, and I'm much less confused now.

Reply:
It's hard to say for sure, since the schema you give doesn't parse.  I think you might mean

`class County
  name ~ string
  state ~ stringcat(strings="...")

class Obs
  county ~ County
  rent ~ real
  room ~ stringcat(strings="...)

observe
  county.name as County
  county.state as State
  room as "Room Type"
  rent as "Monthly Rent"
  from Obs
`

But anyway, the schema used in pclean_lib_test is not quite what make_hirm_schema would output.  Sorry if that caused confusion, but I was optimizing for something simple that would exercise translate_observations.

From the above schema, make_hirm_schema would output four clean_relations named "County:name", "County:state",  "Obs:rent" and "Obs:room", and four noisy_relations named "County", "State", "Room Type" and "Monthly Rent".  Other than the slight name differences, they are basically the same as you have in your expected output.

Reply:
Sorry for writing a non-parsable schema, I just wanted to verify the class definitions, and you're right, I was assuming there was an observation for each attribute. 

Thanks for clarifying what `make_hirm_schema` would output. The Model 5 section of the overleaf seems ambiguous as to whether we should have noisy observations for attributes in $C*$ (as `make_hirm_schema` implements) or whether $C*$ is assumed to contain clean observations (as in the test you wrote). We should ask them to clarify (and fix it in a follow-up if need be, not this PR).